README.md
# Keboola PHP Encryption [![Build Status](https://travis-ci.org/keboola/php-encryption.png?branch=master)](https://travis-ci.org/keboola/php-encryption)

Wrapper for PHP [mcrypt](http://php.net/manual/en/book.mcrypt.php) encryption algorithms.

## Currently supported ciphers
 * AES in CBC mode with [PKCS7 padding](http://en.wikipedia.org/wiki/Padding_(cryptography\)#PKCS7)
  *  Encryption strength is determined by key length. Key lengths 16, 24, 32 corresponds with 128, 192, 256 bit encryption.

## Usage

```php
    use Keboola\Encryption\AesEncryptor;
	$encryptor = new AesEncryptor('UfhZPgPLpz7YVjXwNGTpUD8WpoddfpXn'); // 256 bit key
    $encryptedMessage = $encryptor->encrypt($inputMessage);
```

## Resources
 * https://gist.github.com/RiANOl/1077723
 * http://php.net/manual/en/function.mcrypt-encrypt.php
 * https://en.wikipedia.org/wiki/Block_cipher_mode_of_operation
 * http://www.coderelic.com/2011/10/aes-256-encryption-with-php/
 * http://en.wikipedia.org/wiki/Advanced_Encryption_Standard


CHANGELOG.md
# 0.2.1
 * [Bugfix] Tests fixed

# 0.2.0
 * [BC break] Acronyms in class names uses `StudlyCaps`

# 0.1.1
 * [Bugfix] Use non blocking random source for initialization vector create

# 0.1.0
 * [BC break] Class AES128Encryptor renamed to AESEncryptor
 * [BC break] Key size must be 128, 192 or 256 bits
 * [Refactoring] AESEncryptor switched to advanced mcrypt API


# 0.0.1
 * Initial release

tests/benchmark.php
<?php
/**
 * User: martinhalamicek
 * Date: 7/15/13
 * Time: 10:04 AM
 */

require_once __DIR__ . '/../vendor/autoload.php';

$keys = array(
	'tEhb6pAWGsZ8orFR',
	'tdyrB9JdFaQ8ArYNvKuvyuBg',
	'dTfYyTgiG7TFvoHMVVtJpw4moiRLxpEi',
);


foreach ($keys as $key) {
	echo 'Strength: ' . (strlen($key) * 8) . PHP_EOL;
	$startTime = microtime(true);
	$encryptor = new \Keboola\Encryption\AESEncryptor($key);

	$i = 0;
	do {
		$encryptedData = $encryptor->encrypt(openssl_random_pseudo_bytes(1000));
		$encryptor->decrypt($encryptedData);
	} while($i++ < 100000 );

	echo 'Done in: ' . (microtime(true) - $startTime) . PHP_EOL;
}

tests/Keboola/Encryption/AESEncryptorTest.php
<?php
/**
 * User: martinhalamicek
 * Date: 7/12/13
 * Time: 3:32 PM
 */

namespace Keboola\Encryption\Tests;

use Keboola\Encryption\AESEncryptor;

class AESEncryptorTest extends \PHPUnit_Framework_TestCase
{

	private $key128bit = 'u7Vd3wukDDghgMzZ';

	public function testEncryptedMessageShouldNotBeEqualToSourceMessage()
	{
		$encryptor = new AESEncryptor($this->key128bit);
		$inputMessage = 'someMessage';
		$this->assertNotEquals($inputMessage, $encryptor->encrypt($inputMessage));
	}

	/**
	 * @dataProvider inputMessages
	 * @param $inputMessage
	 */
	public function testEncryptionShouldNotChangeMessage($inputMessage)
	{
		$encryptor = new AESEncryptor($this->key128bit);
		$this->assertEquals($inputMessage, $encryptor->decrypt($encryptor->encrypt($inputMessage)));
	}

	public function inputMessages()
	{
		return array_map(function($item) { return array($item); }, array(
			'some message',
			' message with spaces ',
			"with line

			break \0\0"
		));
	}

	public function testEncryptedMessageIsDifferentForSameInputs()
	{
		$inputMessage = 'someMessage';

		$encryptor = new AESEncryptor($this->key128bit);
		$this->assertNotEquals($encryptor->encrypt($inputMessage), $encryptor->encrypt($inputMessage));
	}

	/**
	 *  @expectedException \InvalidArgumentException
	 */
	public function testExceptionShouldBeThrownOnInvalidKeyLength()
	{
		new AESEncryptor('key');
	}

}

src/Keboola/Encryption/AES128Encryptor.php
<?php
/**
 * AES Encryption with PKCS7 padding http://en.wikipedia.org/wiki/Padding_(cryptography)#PKCS7
 * https://gist.github.com/RiANOl/1077723
 * http://php.net/manual/en/function.mcrypt-encrypt.php
 * http://stackoverflow.com/questions/7448763/proper-php-mcrypt-encryption-methods
 *
 * User: martinhalamicek
 * Date: 7/8/13
 * Time: 3:06 PM
 */

namespace Keboola\Encryption;

class AES128Encryptor implements EncryptorInterface
{

	private $_cipher = MCRYPT_RIJNDAEL_128;
	private $_mode = MCRYPT_MODE_CBC;

	private $_key;

	private $_initializationVectorSize;

	public function __construct($key)
	{
		$this->_key = $key;
		$this->_initializationVectorSize = mcrypt_get_iv_size($this->_cipher, $this->_mode);

		if (strlen($key) > ($keyMaxLength = mcrypt_get_key_size($this->_cipher, $this->_mode))) {
			throw new \InvalidArgumentException("The key length must be less or equal than $keyMaxLength.");
		}
	}

	/**
	 * @param $data
	 * @return string
	 */
	public function encrypt($data)
	{
		$blockSize = mcrypt_get_block_size($this->_cipher, $this->_mode);
		$pad = $blockSize - (strlen($data) % $blockSize);
		$iv = mcrypt_create_iv($this->_initializationVectorSize, MCRYPT_DEV_URANDOM);
		return $iv . mcrypt_encrypt(
			$this->_cipher,
			$this->_key,
			$data . str_repeat(chr($pad), $pad),
			$this->_mode,
			$iv
		);
	}

	/**
	 * @param $encryptedData
	 * @return string
	 */
	public function decrypt($encryptedData)
	{
		$initializationVector = substr($encryptedData, 0, $this->_initializationVectorSize);
		$data =  mcrypt_decrypt(
			$this->_cipher,
			$this->_key,
			substr($encryptedData, $this->_initializationVectorSize),
			$this->_mode,
			$initializationVector
		);
		$pad = ord($data[strlen($data) - 1]);
		return substr($data, 0, -$pad);
	}

}

src/Keboola/Encryption/AESEncryptor.php
<?php
/**
 * AES Encryption in CBC mode with PKCS7 padding
 *
 * User: martinhalamicek
 * Date: 7/8/13
 * Time: 3:06 PM
 */

namespace Keboola\Encryption;

class AESEncryptor implements EncryptorInterface
{

	/**
	 * @var string Encryption key
	 */
	private $key;

	/**
	 * @var int Mcrypt initialization vector size
	 */
	private $initializationVectorSize;

	/**
	 * @var resource mcrypt module resource
	 */
	private $mcryptModule;

	/**
	 * @var int encryption block size
	 */
	private $blockSize;

	/**
	 * @param string $key encryption key should be 16, 24 or 32 characters long form 128, 192, 256 bit encryption
	 */
	public function __construct($key)
	{
		$this->mcryptModule = mcrypt_module_open('rijndael-128', '', 'cbc', '');
		if ($this->mcryptModule === false) {
			throw new \InvalidArgumentException("Unknown algorithm/mode.");
		}

		$keyLength = strlen($key);
		if ($keyLength > ($keyMaxLength = mcrypt_enc_get_key_size($this->mcryptModule))) {
			throw new \InvalidArgumentException("The key length must be less or equal than $keyMaxLength.");
		}
		if (!in_array($keyLength, array(16, 24, 32))) {
			throw new \InvalidArgumentException("Key length must be 16, 24 or 32 bytes for 128, 192, 256 bit encryption.");
		}

		$this->key = $key;

		$this->initializationVectorSize = mcrypt_enc_get_iv_size($this->mcryptModule);
		$this->blockSize = mcrypt_enc_get_block_size($this->mcryptModule);
	}

	/**
	 * @param $data
	 * @return string
	 */
	public function encrypt($data)
	{
		$iv = mcrypt_create_iv($this->initializationVectorSize, MCRYPT_DEV_URANDOM);
		mcrypt_generic_init($this->mcryptModule, $this->key, $iv);
		$encrypted = mcrypt_generic($this->mcryptModule, $this->pad($data));
		mcrypt_generic_deinit($this->mcryptModule);
		return $iv. $encrypted;
	}

	/**
	 * @param $encryptedData
	 * @return string
	 */
	public function decrypt($encryptedData)
	{
		$initializationVector = substr($encryptedData, 0, $this->initializationVectorSize);
		mcrypt_generic_init($this->mcryptModule, $this->key, $initializationVector);
		$decryptedData = mdecrypt_generic($this->mcryptModule, substr($encryptedData, $this->initializationVectorSize));
		mcrypt_generic_deinit($this->mcryptModule);
		return $this->unpad($decryptedData);
	}

	private function pad($data)
	{
		$pad = $this->blockSize - (strlen($data) % $this->blockSize);
		return $data . str_repeat(chr($pad), $pad);
	}

	private function unpad($data)
	{
		$pad = ord($data[strlen($data) - 1]);
		return substr($data, 0, -$pad);
	}

	public function __destruct()
	{
		mcrypt_module_close($this->mcryptModule);
	}

}